core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java

```diff
@@ -237,6 +237,14 @@ public synchronized void startLifespan(Lifespan lifespan, ConnectorPartitionHand
         scheduleGroups.put(lifespan, new ScheduleGroup(partitionHandle));
         whenFinishedOrNewLifespanAdded.set(null);
         whenFinishedOrNewLifespanAdded = SettableFuture.create();
+
+        // Avoid deadlocks by immediately scheduling a task for collecting dynamic filters because:
+        // * there can be task in other stage blocked waiting for the dynamic filters, or
+        // * connector split source for this stage might be blocked waiting the dynamic filters.
+        if (dynamicFilterService.isCollectingTaskNeeded(stageExecution.getStageId().getQueryId(), stageExecution.getFragment())) {
```

Comment:
I'm not sure if scheduling is supposed to happen here. Could you please explain why is it necessary to schedule a task in `startLifespan`? Isn't the `schedule` method going to be called at least once before blocking?

Reply:
It's needed to collect dynamic filters for source stages with broadcast joins. 

> Isn't the schedule method going to be called at least once before blocking?

No. With phased scheduler some stages won't be scheduled before all dependent stages enumerate all splits. In this case there might be deadlock (without this extra task), because DF is collected by donwstream stage.

Reply:
This will effectively schedule a task on `StageScheduler` construction before `SqlQueryScheduler` is fully initialized and scheduling is started. 

At very least the state reporting will be inconsistent. The fact that there are some tasks running while the stage is still in a state `PLANNED` might be very non obvious and potentially error prone. Also it's not clear what is going to happen if a scheduled task fails or finishes before `SqlQueryScheduler` is fully initialized and all the listeners are attached. Even if it works today, it might easily break in the future, as it might not be expected for any tasks to exist before the `SqlQueryScheduler` is initialized and the stage is in state `SCHEDULING`. I would strongly recommend to avoid this route.

It generally feels like it should be responsibility of the `ExecutionSchedule` (`PhasedExecutionSchedule`) to determine what stages should be scheduled in what order. I was going to recommend making the `PhasedExecutionSchedule` aware of dynamic filters but it seems like you are already doing it. Based on the dynamic filter dependencies the `PhasedExecutionSchedule` can trigger scheduling for a source stage that contains broadcast join. Then the `SourcePartitionedScheduler` can rely on the existing mechanism on scheduling a single task when the split enumeration is blocked.

Reply:
I've added `io.trino.execution.scheduler.SourceScheduler#start` which is called from within `io.trino.execution.scheduler.SqlQueryScheduler.PipelinedDistributedStagesScheduler#schedule` for every `SourceScheduler`.